pkg/cmd/repo/autolink/list_test.go

```diff
@@ -0,0 +1,256 @@
+package autolink
+
+import (
+	"bytes"
+	"net/http"
+	"testing"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/cli/cli/v2/internal/browser"
+	"github.com/cli/cli/v2/internal/ghrepo"
+	"github.com/cli/cli/v2/pkg/cmdutil"
+	"github.com/cli/cli/v2/pkg/httpmock"
+	"github.com/cli/cli/v2/pkg/iostreams"
+	"github.com/google/shlex"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewCmdList(t *testing.T) {
+	tests := []struct {
+		name         string
+		input        string
+		output       listOptions
+		wantErr      bool
+		wantExporter bool
+		errMsg       string
+	}{
+		{
+			name:   "no argument",
+			input:  "",
+			output: listOptions{},
+		},
+		{
+			name:   "web flag",
+			input:  "--web",
+			output: listOptions{WebMode: true},
+		},
+		{
+			name:         "json flag",
+			input:        "--json id",
+			output:       listOptions{},
+			wantExporter: true,
+		},
+		{
+			name:    "invalid json flag",
+			input:   "--json invalid",
+			wantErr: true,
+			errMsg: heredoc.Doc(`
+				Unknown JSON field: "invalid"
+				Available fields:
+				  id
+				  isAlphanumeric
+				  keyPrefix
+				  urlTemplate`),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ios, _, _, _ := iostreams.Test()
+			f := &cmdutil.Factory{
+				IOStreams: ios,
+			}
+
+			argv, err := shlex.Split(tt.input)
+			require.NoError(t, err)
+
+			var gotOpts *listOptions
+			cmd := newCmdList(f, func(opts *listOptions) error {
+				gotOpts = opts
+				return nil
+			})
+
+			cmd.SetArgs(argv)
+			cmd.SetIn(&bytes.Buffer{})
+			cmd.SetOut(&bytes.Buffer{})
+			cmd.SetErr(&bytes.Buffer{})
+
+			_, err = cmd.ExecuteC()
+			if tt.wantErr {
+				require.EqualError(t, err, tt.errMsg)
+				return
+			}
+
+			require.NoError(t, err)
+			assert.Equal(t, tt.output.WebMode, gotOpts.WebMode)
+			assert.Equal(t, tt.wantExporter, gotOpts.Exporter != nil)
+		})
+	}
+}
+
+func TestListRun(t *testing.T) {
+	tests := []struct {
+		name       string
+		opts       *listOptions
+		isTTY      bool
+		httpStubs  func(t *testing.T, reg *httpmock.Registry)
+		wantStdout string
+		wantStderr string
+		wantErr    bool
+	}{
+		{
+			name:  "list tty",
+			opts:  &listOptions{},
+			isTTY: true,
+			httpStubs: func(t *testing.T, reg *httpmock.Registry) {
+				reg.Register(
+					httpmock.REST("GET", "repos/OWNER/REPO/autolinks"),
+					httpmock.StringResponse(`[
+						{
+							"id": 1,
+							"key_prefix": "TICKET-",
+							"url_template": "https://example.com/TICKET?query=<num>",
+							"is_alphanumeric": true
+						},
+						{
+							"id": 2,
+							"key_prefix": "STORY-",
+							"url_template": "https://example.com/STORY?id=<num>",
+							"is_alphanumeric": false
+						}
+					]`),
+				)
+			},
```

Comment:
I _really_ dislike this testing pattern as it results in overmocking and hard to maintain test suites. I think, since this is a new subcommand set, we may have an opportunity to explore a better pattern, here.

I'll open a new branch against this PR to demonstrate what I mean.

Reply:
Here's my [PR](https://github.com/hoffm/github-cli/pull/1) demonstrating what I'd prefer to see. It can be merged into this branch. Happy to have any discussion you'd like about the implementation pattern there 🙂

Reply:
Thank you for the PR. I agree it's a much better pattern. With my first contribution, I aimed to repeat extant patterns in the codebase, but I'd like to see this pattern propagated throughout.

Reply:
Is a totally sensible decision @hoffm, with all the original maintainers having moved off the team, we're currently figuring out how we would like to address some of these pain points, and as a new command this looks like interesting ground for trying out some of potential patterns.